src/server.c

```diff
@@ -133,23 +175,48 @@ void serverLogRaw(int level, const char *msg) {
     } else {
         int off;
         struct timeval tv;
-        int role_char;
         pid_t pid = getpid();
         int daylight_active = atomic_load_explicit(&server.daylight_active, memory_order_relaxed);
 
         gettimeofday(&tv, NULL);
         struct tm tm;
         nolocks_localtime(&tm, tv.tv_sec, server.timezone, daylight_active);
-        off = strftime(buf, sizeof(buf), "%d %b %Y %H:%M:%S.", &tm);
-        snprintf(buf + off, sizeof(buf) - off, "%03d", (int)tv.tv_usec / 1000);
+        switch (server.log_timestamp_format) {
+        case LOG_TIMESTAMP_LEGACY:
+            off = strftime(buf, sizeof(buf), "%d %b %Y %H:%M:%S.", &tm);
+            snprintf(buf + off, sizeof(buf) - off, "%03d", (int)tv.tv_usec / 1000);
+            break;
+
+        case LOG_TIMESTAMP_ISO8601:
+            off = strftime(buf, sizeof(buf), "%Y-%m-%dT%H:%M:%S.", &tm);
+            char tzbuf[7];
+            formatTimezone(tzbuf, sizeof(tzbuf), server.timezone, server.daylight_active);
+            snprintf(buf + off, sizeof(buf) - off, "%03d%s", (int)tv.tv_usec / 1000, tzbuf);
+            break;
+
+        case LOG_TIMESTAMP_MILLISECONDS:
+            snprintf(buf, sizeof(buf), "%lld", (long long)tv.tv_sec * 1000 + (long long)tv.tv_usec / 1000);
+            break;
+        }
+        int role_index;
         if (server.sentinel_mode) {
-            role_char = 'X'; /* Sentinel. */
+            role_index = 0; /* Sentinel. */
         } else if (pid != server.pid) {
-            role_char = 'C'; /* RDB / AOF writing child. */
+            role_index = 1; /* RDB / AOF writing child. */
         } else {
-            role_char = (server.primary_host ? 'S' : 'M'); /* replica or Primary. */
+            role_index = (server.primary_host ? 2 : 3); /* Slave or Master. */
+        }
+        switch (server.log_format) {
+        case LOG_FORMAT_LOGFMT:
+            filterInvalidLogfmtChar(msg);
+            fprintf(fp, "pid=%d role=%s timestamp=\"%s\" level=%s message=\"%s\"\n", (int)getpid(), roles[role_index],
+                    buf, verbose_level[level], msg);
```

Comment:
If we can't make `msg` mutable, then, what we need to do is to copy to a temporary buffer. But I think we should avoid that and only do it if there are some forbidden characters.

We can check if there are any forbidden characters first, and then copy to a temp buffer if there are any. It's no problem if this case is less optimal because we expect it to be almost never used, so a heap allocated string is OK.

Something like this:

```suggestion
            char *safemsg = msg;
            if (hasInvalidLogfmtChar(msg)) {
                safemsg = filterInvalidLogfmtChar(msg);
            }
            fprintf(fp, "pid=%d role=%s timestamp=\"%s\" level=%s message=\"%s\"\n", (int)getpid(), roles[role_index],
                    buf, verbose_level[level], safemsg);
            if (safemsg != msg) {
                zfree(safemsg);
            }
```

Reply:
Code updated. 
I couldn't do `char *safemsg = msg;` without compiler warning, so I had to do something circuitous.

Reply:
OK, great. My example was untested. :)